File: tasks/python/task-05-textstats-word-count/repo/textstats.py
"""Simple statistics for a block of text."""


def char_count(text):
    """Return the number of characters, spaces included."""
    return len(text)


def line_count(text):
    """Return the number of lines that hold some text."""
    return sum(1 for line in text.splitlines() if line.strip())


def longest_line(text):
    """Return the longest line. Return an empty string if there is none."""
    longest = ""
    for line in text.splitlines():
        if len(line) > len(longest):
            longest = line
    return longest


def most_common_char(text):
    """Return the most common non-space character, or None."""
    counts = {}
    for char in text:
        if char.isspace():
            continue
        counts[char] = counts.get(char, 0) + 1
    if not counts:
        return None
    return max(counts, key=counts.get)


def summary(text):
    """Return the main counts in one dict."""
    return {
        "chars": char_count(text),
        "lines": line_count(text),
    }

File: tasks/typescript/task-05-textstats-word-count/repo/textstats.ts
/** Simple statistics for a block of text. */

export type Summary = {
  chars: number;
  lines: number;
};

/** Return the number of characters, spaces included. */
export function charCount(text: string): number {
  return text.length;
}

/** Return the number of lines that hold some text. */
export function lineCount(text: string): number {
  return text.split("\n").filter((line) => line.trim() !== "").length;
}

/** Return the longest line. Return an empty string if there is none. */
export function longestLine(text: string): string {
  let longest = "";
  for (const line of text.split("\n")) {
    if (line.length > longest.length) {
      longest = line;
    }
  }
  return longest;
}

/** Return the most common non-space character, or null. */
export function mostCommonChar(text: string): string | null {
  const counts = new Map<string, number>();
  for (const char of text) {
    if (/\s/.test(char)) {
      continue;
    }
    counts.set(char, (counts.get(char) ?? 0) + 1);
  }
  let best: string | null = null;
  let bestCount = 0;
  for (const [char, count] of counts) {
    if (count > bestCount) {
      best = char;
      bestCount = count;
    }
  }
  return best;
}

/** Return the main counts in one object. */
export function summary(text: string): Summary {
  return {
    chars: charCount(text),
    lines: lineCount(text),
  };
}
